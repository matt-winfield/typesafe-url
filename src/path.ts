import { ZodObject, z } from "zod";

export interface PathOptions<
  TParams extends ZodObject<any>,
  TQuery extends ZodObject<any>,
> {
  params?: TParams;
  query?: TQuery;
}

export type PathDefinition<
  TParams extends ZodObject<any>,
  TQuery extends ZodObject<any>,
> = {
  pathTemplate: string;
  options: PathOptions<TParams, TQuery>;
};

export const definePath = <
  TParams extends ZodObject<any>,
  TQuery extends ZodObject<any>,
>(
  pathTemplate: string,
  options: PathOptions<TParams, TQuery>,
): PathDefinition<TParams, TQuery> => {
  return {
    pathTemplate,
    options,
  };
};

export const path = <
  TParams extends ZodObject<any>,
  TQuery extends ZodObject<any>,
>(
  pathDefinition: PathDefinition<TParams, TQuery>,
  params: { params?: z.infer<TParams>; query?: z.infer<TQuery> },
) => {
  const { pathTemplate } = pathDefinition;
  const { params: pathParams, query } = params;

  let output = pathTemplate;

  if (pathParams) {
    const serializedPath = replacePathTemplateParams(pathTemplate, pathParams);
    output = serializedPath;
  }

  if (query) {
    const serializedQuery = serializeQueryParams(query);
    output += `?${serializedQuery}`;
  }

  return output;
};

const replacePathTemplateParams = (
  path: string,
  params: Record<string, any>,
) => {
  return path.replace(/:\w+/g, (match) => {
    const key = match.slice(1);
    return params[key];
  });
};

const serializeQueryParams = (query: Record<string, any>) => {
  return Object.keys(query)
    .map((key) => {
      const value = query[key];
      if (typeof value === "object") {
        return serializeObject(value, key);
      }
      return `${key}=${value}`;
    })
    .join("&");
};

const serializeObject = (obj: Record<string, any>, key: string): string => {
  return Object.keys(obj)
    .map((nestedKey) => {
      if (typeof obj[nestedKey] === "object") {
        return serializeObject(obj[nestedKey], `${key}[${nestedKey}]`);
      }
      return `${key}[${nestedKey}]=${obj[nestedKey]}`;
    })
    .join("&");
};
