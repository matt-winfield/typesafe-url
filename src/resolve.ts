import { ZodNumber, ZodObject, z } from "zod";
import { PathDefinition } from "./path";

/**
 * Resolves the path and query parameters from a URL, pathname and/or search string
 * @param pathDefinition The path pathDefinition
 * @param url The URL to resolve
 * @returns The resolved path parameters or null if the URL does not match the path
 * @throws If the path parameters are invalid (e.g. the type does not match the schema)
 * @example
 * const pathDefinition = definePath("/test/:id", {
 *   params: z.object({
 *     id: z.number(),
 *   }),
 * });
 * const result = resolvePathParams(pathDefinition, "/test/123");
 * console.log(result); // { id: 123 }
 */
export const resolve = <
  TParams extends ZodObject<any>,
  TQuery extends ZodObject<any>,
>(
  pathDefinition: PathDefinition<TParams, TQuery>,
  url: string,
) => {
  const pathParams = resolvePathParams(pathDefinition, url);
  const queryParams = resolveQueryParams(pathDefinition, url);

  return {
    params: pathParams,
    query: queryParams,
  };
};

/**
 * Resolves the path parameters from a URL, pathname and/or search string
 * @param pathDefinition The path pathDefinition
 * @param url The URL to resolve
 * @returns The resolved path parameters or null if the URL does not match the path
 * @throws If the path parameters are invalid (e.g. the type does not match the schema)
 * @example
 * const pathDefinition = definePath("/test/:id", {
 *   params: z.object({
 *     id: z.number(),
 *   }),
 * });
 * const result = resolvePathParams(pathDefinition, "/test/123");
 * console.log(result); // { id: 123 }
 */
export const resolvePathParams = <TParams extends ZodObject<any>>(
  pathDefinition: PathDefinition<TParams, ZodObject<any>>,
  url: string,
) => {
  if (!pathDefinition.options.params) {
    return undefined;
  }

  const { pathTemplate } = pathDefinition;

  const path = extractPath(url);

  const pathSegments = pathTemplate.toLowerCase().split("/");
  const urlSegments = path.toLowerCase().split("/");

  const pathParams: Record<string, unknown> = {};

  for (let i = 0; i < pathSegments.length; i++) {
    const pathSegment = pathSegments[i];
    const urlSegment = urlSegments[i];

    if (pathSegment.startsWith(":")) {
      const paramName = pathSegment.slice(1);

      if (!urlSegment) {
        return null;
      }

      const parsedParam = coerceIntoZodType(
        urlSegment,
        pathDefinition.options.params.shape[paramName],
      );
      pathParams[paramName] = parsedParam;
    } else if (pathSegment !== urlSegment) {
      return null;
    }
  }

  return pathDefinition.options.params.parse(pathParams);
};

/**
 * Resolves the query parameters from a URL or search string
 * @param pathDefinition The path pathDefinition
 * @param url The URL to resolve
 * @returns The resolved query parameters
 * @throws If the query parameters are do not match the schema
 * @example
 * const pathDefinition = definePath("/test?value=123", {
 *   query: z.object({
 *     value: z.number(),
 *   }),
 * });
 * const result = resolvePathParams(pathDefinition, "/test?value=123");
 * console.log(result); // { value: 123 }
 */
export const resolveQueryParams = <TParams extends ZodObject<any>>(
  pathDefinition: PathDefinition<TParams, ZodObject<any>>,
  url: string,
) => {
  if (!pathDefinition.options.query) {
    return undefined;
  }
  const query = extractQuery(url);

  let result: Record<string, any> = {};
  for (const [key, value] of query.entries()) {
    // example deeply nested path: "key[test][0][value]=123"
    const path = key.split(/[\[\]]/).filter(Boolean);
    let current = result;
    for (const nestedKey of path.slice(0, -1)) {
      if (!current[nestedKey]) {
        current[nestedKey] = {};
      }
      current = current[nestedKey];
    }

    const shape = getDeeplyNestedValue(
      pathDefinition.options.query.shape,
      path,
    );
    current[path[path.length - 1]] = coerceIntoZodType(value, shape);
  }
  return pathDefinition.options.query.parse(result);
};

const extractPath = (url: string) => {
  if (url.startsWith("/")) {
    return url;
  }

  try {
    const urlObject = new URL(url);
    return urlObject.pathname;
  } catch {
    return "";
  }
};

const extractQuery = (url: string) => {
  try {
    const urlObject = new URL(url);
    return new URLSearchParams(urlObject.search);
  } catch {
    return new URLSearchParams(url);
  }
};

const getDeeplyNestedValue = (obj: any, path: string[]) => {
  let current = obj;
  for (const key of path) {
    if (current[key] instanceof ZodObject) {
      current = current[key].shape;
    } else {
      current = current[key];
    }
  }
  return current;
};

const coerceIntoZodType = (value: any, zodType: any) => {
  if (zodType instanceof ZodNumber) {
    // if the value is a valid number, return it as a number
    if (!isNaN(Number(value))) {
      return Number(value);
    }
    return value;
  }

  return value;
};
