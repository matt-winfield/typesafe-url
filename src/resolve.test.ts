import { z } from "zod";
import { definePath } from "./path";
import { resolve } from "./resolve";

describe("resolve", () => {
  it("should resolve path params", () => {
    const testPath = definePath("/test/:id", {
      params: z.object({
        id: z.number(),
      }),
    });

    const result = resolve(testPath, "/test/123");

    expect(result).toEqual({
      params: {
        id: 123,
      },
    });
  });

  it("should resolve path and query params when full URL is given with string", () => {
    const testPath = definePath("/test/:id", {
      params: z.object({
        id: z.string(),
      }),
      query: z.object({
        test: z.string(),
      }),
    });

    const result = resolve(testPath, "https://example.com/test/123?test=456");

    expect(result).toEqual({
      params: {
        id: "123",
      },
      query: {
        test: "456",
      },
    });
  });

  it("should resolve path and query params when full URL is given with number", () => {
    const testPath = definePath("/test/:id", {
      params: z.object({
        id: z.number(),
      }),
      query: z.object({
        test: z.number(),
      }),
    });

    const result = resolve(testPath, "https://example.com/test/123?test=456");

    expect(result).toEqual({
      params: {
        id: 123,
      },
      query: {
        test: 456,
      },
    });
  });

  it("should throw an error if the path type doesn't match", () => {
    const testPath = definePath("/test/:id/:test", {
      params: z.object({
        id: z.number(),
        test: z.number(),
      }),
    });

    expect(() => {
      resolve(testPath, "https://example.com/test/something/else");
    }).toThrowErrorMatchingSnapshot();
  });

  it("should throw an error if the query type doesn't match", () => {
    const testPath = definePath("/test", {
      query: z.object({
        test1: z.number(),
        test2: z.number(),
      }),
    });

    expect(() => {
      resolve(testPath, "https://example.com/test?test1=blah&test2=blah");
    }).toThrowErrorMatchingSnapshot();
  });

  it("should resolve deeply nested objects in query params", () => {
    const testPath = definePath("/test", {
      query: z.object({
        nested: z.object({
          deeply: z.object({
            deeper: z.object({
              deepest: z.string(),
            }),
          }),
          test: z.object({
            nested: z.number(),
          }),
        }),
      }),
    });

    const result = resolve(
      testPath,
      "https://example.com/test?nested[deeply][deeper][deepest]=nested&nested[test][nested]=123",
    );

    expect(result).toEqual({
      query: {
        nested: {
          deeply: {
            deeper: {
              deepest: "nested",
            },
          },
          test: {
            nested: 123,
          },
        },
      },
    });
  });
});
