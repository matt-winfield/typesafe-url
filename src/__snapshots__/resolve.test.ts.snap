// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`resolve > should throw an error if the path type doesn't match 1`] = `
[ZodError: [
  {
    "code": "invalid_type",
    "expected": "number",
    "received": "string",
    "path": [
      "id"
    ],
    "message": "Expected number, received string"
  },
  {
    "code": "invalid_type",
    "expected": "number",
    "received": "string",
    "path": [
      "test"
    ],
    "message": "Expected number, received string"
  }
]]
`;

exports[`resolve > should throw an error if the query type doesn't match 1`] = `
[ZodError: [
  {
    "code": "invalid_type",
    "expected": "number",
    "received": "string",
    "path": [
      "test1"
    ],
    "message": "Expected number, received string"
  },
  {
    "code": "invalid_type",
    "expected": "number",
    "received": "string",
    "path": [
      "test2"
    ],
    "message": "Expected number, received string"
  }
]]
`;
